ast/compile.go

```diff
@@ -4761,7 +4761,12 @@ func checkUnusedAssignedVars(loc *Location, stack *localDeclaredVars, used VarSe
 	unused = unused.Diff(rewrittenUsed)
 
 	for _, gv := range unused.Sorted() {
-		errs = append(errs, NewError(CompileErr, loc, "assigned var %v unused", dvs.reverse[gv]))
+		for i := range body {
+			if body[i].Vars(VarVisitorParams{}).Contains(gv) {
```

Comment:
To limit what might end up in this comparison, consider implementing an `assignedVars()`, akin to 

https://github.com/open-policy-agent/opa/blob/d166d11b961e09343b277a0f485aa304db50c593/ast/compile.go#L4119

Where we only consider vars in `*Expr` types that meet the `x.IsAssignment() && validEqAssignArgCount(x)` requirement.

Reply:
hey @johanfylling thanks for the review! 
maybe a newbie question
I've tried implementing `assignedVars()` as you described. The problem is although I can catch the type as *Expr in the switch-case,
all the `x.IsAssigment()` calls return `false` and `validEqAssignArgCount(x)` calls return `true` against the example policy I used in the desc
e.g `for expr:  __local3__ = 3 is assignment:  false validEqAssignArgCount:  true` for unassigned `c`

any idea where I made a mistake?

Reply:
if I remove the last two checks, it works well (w/o using x.IsAssigment() and validEqAssignArgCount control)

```go
func assignedVars(x interface{}) VarSet {
	vars := NewVarSet()
	vis := NewGenericVisitor(func(x interface{}) bool {
		switch x := x.(type) {
		case *Expr:
			WalkVars(x.Operand(0), func(v Var) bool {
				vars.Add(v)
				return false
			})
		}
		return false
	})
	vis.Walk(x)
	return vars
}
```
and
```go
assignedVars := assignedVars(body[i])
if assignedVars.Contains(gv) {
```

Reply:
You can disregard my suggestion; looks like the `assign` expression has already been rewritten to an `eq` at this point, so the filtering I had in mind can't happen.

Reply:
ack, thanks for the feedback! I'll resolve this and move on to the other ones to implement

Reply:
👍 . Sorry for leading you down the wrong path there.